npk/src/archive.rs

```diff
@@ -190,83 +192,52 @@ impl TryFrom<SerdeHashes> for Hashes {
     }
 }
 
-const MANIFEST: &str = "manifest.yaml";
-const SIGNATURE: &str = "signature.yaml";
-const FS_IMAGE: &str = "fs.img";
-
-pub struct ArchiveReader<'a> {
+pub struct ArchiveReader {
     archive: zip::ZipArchive<std::io::BufReader<std::fs::File>>,
-    signing_keys: &'a HashMap<String, ed25519_dalek::PublicKey>,
-}
-
-pub fn read_manifest(
-    npk: &Path,
-    signing_keys: &HashMap<String, ed25519_dalek::PublicKey>,
-) -> Result<Manifest, Error> {
-    let mut archive_reader = ArchiveReader::new(&npk, &signing_keys)?;
-    archive_reader.extract_manifest_from_archive()
 }
 
-impl<'a> ArchiveReader<'a> {
-    pub fn new(
-        npk: &Path,
-        signing_keys: &'a HashMap<String, ed25519_dalek::PublicKey>,
-    ) -> Result<Self, Error> {
+impl ArchiveReader {
+    pub fn new(npk: &Path) -> Result<Self, Error> {
         let file =
             std::fs::File::open(&npk).map_err(|_e| Error::CouldNotOpenFile(PathBuf::from(npk)))?;
-
         let reader: std::io::BufReader<std::fs::File> = std::io::BufReader::new(file);
         let archive: zip::ZipArchive<std::io::BufReader<std::fs::File>> =
             zip::ZipArchive::new(reader).map_err(Error::Zip)?;
-        Ok(Self {
-            archive,
-            signing_keys,
-        })
+        Ok(Self { archive })
     }
 
     pub fn extract_fs_start_and_size(&mut self) -> Result<(u64, u64), Error> {
-        let f = self
+        let zip_file = self
             .archive
             .by_name(FS_IMAGE)
             .map_err(|e| Error::ArchiveError(format!("Failed to find file-system {}", e)))?;
 
-        Ok((f.data_start(), f.size()))
+        Ok((zip_file.data_start(), zip_file.size()))
     }
 
-    pub fn extract_hashes(&mut self) -> Result<Hashes, Error> {
-        let mut signature_file = self
-            .archive
-            .by_name(SIGNATURE)
-            // .with_context(|| "Failed to read signature".to_string())
-            .map_err(Error::Zip)?;
+    pub fn extract_hashes(
```

Comment:
We should make the `ArchiveReader` functions async. To do that, take a look at https://docs.rs/tokio/0.3.5/tokio/task/index.html#blocking-and-yielding
this allows the code in an async function to call a blocking API

Reply:
Can't we make the npk crate just operate on a `std::io::Read`? This avoid the dependency on `tokio` in the `npk` crate. The wrapping in a `spawn_blocking` etc call is up to the user and the usage of tokio should not be dictated but this crate.

```rust
pub struct Npk<'a, R: io::Read + io::Seek> {
    archive: zip::ZipArchive<io::BufReader<R>>,
    signing_keys: &'a HashMap<String, ed25519_dalek::PublicKey>,
}

impl<'a, R: io::Read + io::Seek> Npk<'a, R> {
    pub fn new(
        npk: R,
        signing_keys: &'a HashMap<String, ed25519_dalek::PublicKey>,
    ) -> Result<Self, Error> {
        let reader = io::BufReader::new(npk);
        let archive = zip::ZipArchive::new(reader).map_err(Error::Zip)?;
        Ok(Self {
            archive,
            signing_keys,
        })
    }
}
```

Reply:
yes that of course is true...keep forgetting that we live in the tokio universe and not in the rust universe here :(

Reply:
Why?

Reply:
let's do that in another pr